### PR TITLE
Fix penalty-based cost function in STOMP

### LINKED ISSUE #2462: Random crash with STOMP with path constraints
### Description
There is a random SIGSEGV when using STOMP with path constraints in my project.

Here is the backtrace
> #0  0x0000ffff9a708298 in ?? () from /lib/aarch64-linux-gnu/libfcl.so.0.7
#1  0x0000ffff9a7085c0 in ?? () from /lib/aarch64-linux-gnu/libfcl.so.0.7
#2  0x0000ffff9a708658 in ?? () from /lib/aarch64-linux-gnu/libfcl.so.0.7
#3  0x0000ffff9a708658 in ?? () from /lib/aarch64-linux-gnu/libfcl.so.0.7
#4  0x0000ffff9a704f70 in fcl::DynamicAABBTreeCollisionManager<double>::registerObjects(std::vector<fcl::CollisionObject<double>*, std::allocator<fcl::CollisionObject<double>*> > const&) ()
   from /lib/aarch64-linux-gnu/libfcl.so.0.7
#5  0x0000ffff9acf0094 in collision_detection::FCLObject::registerTo (
    this=this@entry=0xffff4de40be8, manager=0xffff38009880)
    at /.../src/external/ros-planning/moveit2/moveit_core/collision_detection_fcl/src/collision_common.cpp:997
#6  0x0000ffff9acfe0ec in collision_detection::CollisionEnvFCL::allocSelfCollisionBroadPhase (
    this=this@entry=0xffff380237f0, state=..., manager=...)
    at /.../src/external/ros-planning/moveit2/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:254
#7  0x0000ffff9acfe278 in collision_detection::CollisionEnvFCL::checkSelfCollisionHelper (
    this=0xffff380237f0, req=..., res=..., state=..., acm=0xffff38009970)
    at /.../src/external/ros-planning/moveit2/moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:274
#8  0x0000ffff9b2d30a8 in planning_scene::PlanningScene::isStateColliding (this=0xffff38018840, 
    state=..., group=..., verbose=<optimized out>)
    at /.../src/external/ros-planning/moveit2/moveit_core/planning_scene/src/planning_scene.cpp:2049
#9  0x0000ffff4feee88c in stomp_moveit::get_positions (state=..., 
    joints=std::vector of length -155488448280990176, capacity -155664653690366117 = {...})
    at /.../src/external/ros-planning/moveit2/moveit_planners/stomp/include/stomp_moveit/conversion_functions.hpp:59

I may have found the root cause after some digging.

https://github.com/ros-planning/moveit2/blob/9249e44330d241559ed868dbc293fc2a3d1e1907/moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp#L124-L128
here when `invalid_windows.back().first` is `0` and `timestep` is `1`

https://github.com/ros-planning/moveit2/blob/9249e44330d241559ed868dbc293fc2a3d1e1907/moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp#L139-L147
`mu = 0.5 * (start + end)` will be `0` and `costs(j)` became `inf` which is invalid for later use and finally cause the crash.

I can workaround it but don't know how to fix it properly.
please let me know if I'm wrong, thanks.

### Your environment
* ROS Distro: Humble
* OS Version:  Ubuntu 22.04
* Source build: tags/2.8.0
* Cyclone DDS

### PR DESCRIPTION
**Pull request by henningkayser**
This adds several test cases for STOMP's noise generation and cost functions, and provides the following fixes:

* out-of-bounds vector access when tail states of trajectory are invalid
* smoothed costs overriding values of previous invalid groups
* missing validity check of last state in trajectory
* inability to disable cost function interpolation steps
* total cost of trajectory not summing up to sum of state penalties
* bug in Gaussian producing infinite values with invalid start states

Fixes #2462, #2554

### CODE REVIEW

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp [hunk 1]
```diff
@@ -86,46 +86,60 @@ CostFn getCostFunctionFromStateValidator(const StateValidatorFn& state_validator
...
-        costs(timestep + 1) = interpolation_fraction * penalty;
+        // Mark solution as invalid
         validity = false;
 
-        // Open new invalid window when this is the first detected invalid state in a group
+        // OPEN new invalid window when this is the first detected invalid state in a group
         if (!in_invalid_window)
         {
-          invalid_windows.emplace_back(timestep, values.cols());
+          invalid_windows.emplace_back(timestep, timestep);
```
Comment (sjahr) [inline moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:132]: Why do we add current state to the current state?

Comment (sjahr) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:132]: Can you add a descriptive comment for `invalid_windows` when it is initialized earlier in the code?

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp [hunk 2]
```diff
@@ -86,46 +86,60 @@ CostFn getCostFunctionFromStateValidator(const StateValidatorFn& state_validator
     // Subsequent invalid states are assumed to have the same cause, so we are keeping track
     // of "invalid windows" which are used for smoothing out the costs per violation cause
     // with a gaussian, penalizing neighboring valid states as well.
-    for (int timestep = 0; timestep < values.cols() - 1; ++timestep)
+    for (int timestep = 0; timestep < values.cols(); ++timestep)
```
Comment (sjahr) [inline moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:89]: Should timestep be renamed to something like sample_way-point?

File: moveit_planners/stomp/test/test_cost_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  auto cost_fn =
+      stomp_moveit::costs::getCostFunctionFromStateValidator(state_validator_fn, 0.1 /* interpolation_step_size */);
+
+  // Build trajectory with TIMESTEPS states, interpolating from 0.0 to 1.0 joint values
+  Eigen::MatrixXd values = Eigen::MatrixXd::Zero(VARIABLES, TIMESTEPS);
+  const int last_timestep = values.cols() - 1;
+  for (int timestep = 0; timestep <= last_timestep; ++timestep)
+  {
+    values.col(timestep).fill(1.0 * timestep / last_timestep);
```
Comment (sjahr) [inline moveit_planners/stomp/test/test_cost_functions.cpp:59]: What's the purpose of multiplying the fraction with 1.0?

File: moveit_planners/stomp/test/test_cost_functions.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ */
+
+#include <gtest/gtest.h>
+#include <stomp_moveit/cost_functions.hpp>
+
+constexpr size_t TIMESTEPS = 100;
+constexpr size_t VARIABLES = 6;
+constexpr double PENALTY = 1.0;
+
+TEST(NoiseGeneratorTest, testGetCostFunctionAllValidStates)
```
Comment (sjahr) [inline moveit_planners/stomp/test/test_cost_functions.cpp:46]: Can you add a brief description for the unit-test? I like the GIVEN-WHEN-THEN notation

**Review by sjahr — COMMENTED** [review]
Thanks for fixing this! I have a couple of comments, mostly question to understand the changes better. Is there a good way to test the improvements?

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp (hunk 1 above)
Comment (henningkayser) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:132]: What is missing from the description at the top of the loop?

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp (hunk 2 above)
Comment (henningkayser) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:89]: It's really just the timestep in the STOMP trajectory, though. It doesn't contain joint values

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp [hunk 5]
```diff
@@ -86,46 +86,60 @@ CostFn getCostFunctionFromStateValidator(const StateValidatorFn& state_validator
...
+      bool found_invalid_state = costs(timestep) > 0.0;
+
+      // If state is valid, interpolate towards the next waypoint if there is one
+      bool continue_interpolation =
+          !found_invalid_state && timestep < (values.cols() - 1) && interpolation_step_size > 0.0;
+      if (continue_interpolation)
       {
-        Eigen::VectorXd sample_vec = (1 - interpolation_fraction) * current + interpolation_fraction * next;
+        Eigen::VectorXd next = values.col(timestep + 1);
+        const double interpolation_step = std::min(0.5, interpolation_step_size / (next - current).norm());
```
[the remark this answers is not in the document]
Comment (henningkayser) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:105]: plus, this is a filter for enforcing interpolation even if waypoints are really close to each other.

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp (hunk 1 above)
Comment (henningkayser) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:136]: I added more remarks to the documentation

File: moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp [hunk 6]
```diff
@@ -86,46 +86,60 @@ CostFn getCostFunctionFromStateValidator(const StateValidatorFn& state_validator
     // Subsequent invalid states are assumed to have the same cause, so we are keeping track
     // of "invalid windows" which are used for smoothing out the costs per violation cause
     // with a gaussian, penalizing neighboring valid states as well.
-    for (int timestep = 0; timestep < values.cols() - 1; ++timestep)
+    for (int timestep = 0; timestep < values.cols(); ++timestep)
     {
+      // Get state at current timestep and check for validity
+      // The penalty of the validation function is added to the cost of the current timestep
+      // A state is rendered invalid if a cost results from the current validity check or if a penalty is carried over
```
[the remark this answers is not in the document]
Comment (henningkayser) [reply moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp:95]: No cost thresholds supported, currently. They would have to be implemented with the validator function.

**Review by sjahr — APPROVED** [review]

**Pull request merged**